langkit/lkt_lowering.py

```diff
@@ -1360,35 +1366,81 @@ def lower_base_field(
         assert isinstance(decl, L.FieldDecl)
         annotations = parse_annotations(self.ctx, FieldAnnotations, full_decl)
         field_type = self.resolve_type_decl(decl.f_decl_type.p_designated_type)
+        doc = self.ctx.lkt_doc(full_decl)
 
-        cls: Type[BaseField]
-        kwargs: Dict[str, Any]
-        default_value = (self.lower_expr(decl.f_default_val, {})
-                         if decl.f_default_val else None)
+        cls: Type[AbstractNodeData]
+        constructor: Callable[..., AbstractNodeData]
+        kwargs: Dict[str, Any] = {'type': field_type, 'doc': doc}
+
+        if annotations.lazy:
+            check_source_language(
+                not annotations.null_field,
+                'Lazy fields cannot be null'
+            )
+            cls = PropertyDef
+            constructor = create_lazy_field
+
+            _, expr = self.lower_property_expr(
+                abstract=annotations.abstract,
+                external=False,
+                arg_decl_list=None,
+                body=decl.f_default_val,
+            )
+
+            kwargs = {
+                'expr': expr,
+                'doc': doc,
+                'public': annotations.export,
+                'return_type': field_type,
+                'kind': (AbstractKind.abstract
+                         if annotations.abstract
+                         else AbstractKind.concrete),
+            }
+
+        elif annotations.parse_field:
+            assert decl.f_default_val is None
+            check_source_language(
```

Comment:
I'm a bit uncomfortable with those checks here: We have most of the legality checks in `parser.py`, wouldn't it be better to generally put those there ?

Reply:
As discussed live: I’ll turn these into asserts with TODOs to add the corresponding checks in liblktlang. However I would prefer to implement these TODOs in another PR.